pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,464 @@
+		return err
+	}
+
+	dnsEndpointObject, err := d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			newDNSEndpointObject := &feddnsv1a1.DNSEndpoint{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: namespace,
+				},
+				Spec: feddnsv1a1.DNSEndpointSpec{
+					Endpoints: dnsEndpoints,
+				},
+			}
+
+			_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Create(newDNSEndpointObject)
+		}
+		return err
+	}
+
+	dnsEndpointObject.Spec.Endpoints = dnsEndpoints
+	_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Update(dnsEndpointObject)
+	return err
+}
+
+// GetEndpointsForServiceDNSObject returns endpoint objects for each MultiClusterServiceDNSRecord object that should be processed.
+func GetEndpointsForServiceDNSObject(dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord) ([]*feddnsv1a1.Endpoint, error) {
+	endpoints, err := getNamesForDNSObject(dnsObject, dnsObject.Spec.FederationName, dnsObject.Spec.DNSSuffix)
+	if err != nil {
+		return nil, err
+	}
+
+	return DedupeAndMergeEndpoints(endpoints), nil
+}
+
+// Merge and remove duplicate endpoints
+func DedupeAndMergeEndpoints(endpoints []*feddnsv1a1.Endpoint) (result []*feddnsv1a1.Endpoint) {
+	// Sort endpoints by DNSName
+	sort.Slice(endpoints, func(i, j int) bool {
+		return endpoints[i].DNSName < endpoints[j].DNSName
+	})
+
+	// Remove the endpoint with no targets/ empty targets
+	for i := 0; i < len(endpoints); {
+		for j := 0; j < len(endpoints[i].Targets); {
+			if endpoints[i].Targets[j] == "" {
+				endpoints[i].Targets = append(endpoints[i].Targets[:j], endpoints[i].Targets[j+1:]...)
+				continue
+			}
+			j++
+		}
+		if len(endpoints[i].Targets) == 0 {
+			endpoints = append(endpoints[:i], endpoints[i+1:]...)
+			continue
+		}
+		i++
+	}
+
+	// Merge endpoints with same DNSName
+	for i := 1; i < len(endpoints); {
+		if endpoints[i].DNSName == endpoints[i-1].DNSName {
+			// Merge targets
+			endpoints[i-1].Targets = append(endpoints[i-1].Targets, endpoints[i].Targets...)
+			endpoints[i-1].Targets = sortAndRemoveDuplicateTargets(endpoints[i-1].Targets)
+
+			// Remove the duplicate endpoint
+			endpoints = append(endpoints[:i], endpoints[i+1:]...)
+			continue
+		}
+		i++
+	}
+
+	return endpoints
+}
+
+func sortAndRemoveDuplicateTargets(targets []string) []string {
+	sort.Slice(targets, func(i, j int) bool {
+		return targets[i] < targets[j]
+	})
+	for i := 1; i < len(targets); {
+		if targets[i] == targets[i-1] {
+			// Remove duplicate target
+			targets = append(targets[:i], targets[i+1:]...)
+			continue
+		}
+		i++
+	}
+	return targets
+}
+
+func getNamesForDNSObject(dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord, federation, dnsSuffix string) ([]*feddnsv1a1.Endpoint, error) {
+	endpoints := []*feddnsv1a1.Endpoint{}
+
+	commonPrefix := strings.Join([]string{dnsObject.Name, dnsObject.Namespace, federation, "svc"}, ".")
```

Comment:
(No action required) Consider extracting construction of the common prefix into a utility function since it appears once here and 9 times in the tests.